### PR TITLE
Protect retention from overflowing

### LINKED ISSUE #5111: Overflow in storage.tsdb.max-block-duration
## Bug Report
**What did you do?**

Set `--storage.tsdb.retention=500y`

**What did you expect to see?**

`storage.tsdb.max-block-duration` in /flags would be "50y"

**What did you see instead? Under which circumstances?**

`storage.tsdb.max-block-duration` returns `-267874407370ms`

**Environment**

* System information:

`Linux 4.18.0-13-generic x86_64`

* Prometheus version:

```
prometheus, version 2.6.0 (branch: HEAD, revision: dbd1d58c894775c0788470944b818cc724f550fb)
  build user:       root@bf5760470f13
  build date:       20181217-15:14:46
  go version:       go1.11.3
```

### PR DESCRIPTION
**Pull request by gouthamve**
Also sanitise the max block duration to max a week.

Fixes #5111

### CODE REVIEW

File: cmd/prometheus/main.go [hunk 1]
```diff
@@ -244,8 +245,22 @@ func main() {
...
 	if cfg.tsdb.MaxBlockDuration == 0 {
 		cfg.tsdb.MaxBlockDuration = cfg.tsdb.Retention / 10
+
+		weekLong, err := model.ParseDuration("7d")
+		if err != nil {
+			panic(err)
+		}
+
+		if cfg.tsdb.MaxBlockDuration > weekLong {
+			cfg.tsdb.MaxBlockDuration = weekLong
```
Comment (brian-brazil) [inline cmd/prometheus/main.go]: How was this value determined? I'd be looking at something in the one month to one year range.

Comment (brian-brazil) [reply cmd/prometheus/main.go]: That was before the memory improvements I made in the last release. Having too small blocks will actually increase memory usage.

Comment (SuperQ) [reply cmd/prometheus/main.go]: I still see 20% bump in memory compacting 7d blocks on a typical server with 100k samples/sec.

What is the the overhead of having more blocks?

Comment (brian-brazil) [reply cmd/prometheus/main.go]: It's duplicating the index memory structures, plus the various other bits and pieces. It'll also slow down and require more memory for queries across blocks.

**Comment by gouthamve** [comment]
@brian-brazil PTAL, made it a month.

File: cmd/prometheus/main.go (hunk 1 above)
Comment (SuperQ) [reply cmd/prometheus/main.go]: Right, but how much is that in terms of bytes or percent of server? I see a nearly linear percent burst based on the size of the compaction, so if we went with 28d blocks on my server, I would see an 80% burst in memory use on compaction.

**Comment by gouthamve** [comment]
Nope, those blocks that are already large would be ignored. Should break anything.

File: cmd/prometheus/main.go (hunk 1 above)
Comment (brian-brazil) [reply cmd/prometheus/main.go]: With the current code, I'd expect it to always be around 10% as it'd be creating an index for up to 10% of data.

**Comment by gouthamve** [comment]
Should not* :P edited the original comment also.

**Comment by simonpasquier** [comment]
:+1: 

I hope it isn't a Freudian slip :smirk:

**Pull request merged**